### PR TITLE
feat: implement @llm.prompt decorator (no context or template)

### PR DESCRIPTION
**Pull request by teamdandelion**

### CODE REVIEW

**Comment by teamdandelion** [comment]
* **#1248** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1248?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1247** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1246** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1246?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1244** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1244?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1245** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1245?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1243** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1243?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1241** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1240** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1240?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1242** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1242?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1242?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1239** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1239?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `v2`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python/mirascope/llm/prompts/decorator.py [hunk 1]
```diff
@@ -1,138 +1,173 @@
...
     MessagesPrompt,
+    Prompt,
 )
 
 
-class PromptDecorator(Protocol[DepsT]):
+def _is_messages(
+    messages_or_content: list[Message] | UserContent,
+) -> TypeIs[list[Message]]:
+    return isinstance(messages_or_content, list) and all(
```
Comment (willbakst) [inline python/mirascope/llm/prompts/decorator.py]: I think we can get away with cheating here by only checking the first element. If the first element is a message type, we can safely assume the type is `list[Message]`, right?

Something like this:

```python
return (
    messages_or_content
    and isinstance(messages_or_content, list)
    and isinstance(
        messages_or_content[0], SystemMessage | UserMessage | AssistantMessage
    )
)
```

Comment (teamdandelion) [reply python/mirascope/llm/prompts/decorator.py]: Ok, I updated it (a little more complicated than your version). I also added some tests documenting the pathological edge cases that come up if someone doesn't follow the type hints.

File: python/tests/llm/prompts/test_promt_decorator.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            return bad_contents
+
+        assert bad_prompt() == snapshot(
+            [
+                llm.UserMessage(
+                    content=[
+                        llm.Text(text="so uncivilized"),
+                        llm.UserMessage(content=[llm.Text(text="hello there")]),  # pyright: ignore[reportArgumentType]
+                    ]
+                )
```
Comment (teamdandelion) [inline python/tests/llm/prompts/test_promt_decorator.py]: Considering whether we want to defensively program `llm.messages.user` to throw a TypeError in this case, although it would have a small negative runtime perf impact of checking all the content to make sure it's not a message. We could also put a flag on Mirascope messages like `_is_mirascope_message: True` and make those checks (and the checks here) more efficient

Comment (willbakst) [reply python/tests/llm/prompts/test_promt_decorator.py]: We don't need to check this at all because we are relying on the type system. User error is user error.

This test seems like it should exist in `typechecking` not here.

File: python/tests/llm/prompts/test_promt_decorator.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        """Test async prompt decorator with empty return value."""
+
+        @llm.prompt()
+        async def empty_prompt() -> llm.messages.UserContent:
+            return []
+
+        with pytest.raises(ValueError, match="Prompt returned empty content"):
+            await empty_prompt()
+
+    async def test_async_prompt_with_empty_string(
```
Comment (willbakst) [inline python/tests/llm/prompts/test_promt_decorator.py]: Do we need both sync and async tests for this? I guess that's better for completeness?

Regardless, I would separate the empty value error tests into their own class like `TestValueError` or `TestEmptyPrompt`?

File: python/mirascope/llm/prompts/decorator.py
Comment (willbakst) [inline python/mirascope/llm/prompts/decorator.py:1]: Here is a reimplementation of the decorators to enable `llm.prompt` and `llm.context_prompt` with no arguments so that my up stack comment can be resolved:

```python
"""The `prompt` decorator for writing messages as string templates."""

import inspect
from collections.abc import Awaitable, Callable
from typing import Concatenate, Generic, Protocol, TypeAlias, overload

from typing_extensions import TypeIs

from ..context import Context, DepsT
from ..messages import (
    AssistantMessage,
    Message,
    SystemMessage,
    UserContent,
    UserMessage,
    user,
)
from ..types import P
from .types import (
    AsyncContextMessagesPrompt,
    AsyncContextPrompt,
    AsyncMessagesPrompt,
    AsyncPrompt,
    ContextMessagesPrompt,
    ContextPrompt,
    MessagesPrompt,
    Prompt,
)


def _is_messages(
    messages_or_content: list[Message] | UserContent,
) -> TypeIs[list[Message]]:
    if not messages_or_content:
        raise ValueError("Prompt returned empty content")
    return isinstance(messages_or_content, list) and isinstance(
        messages_or_content[0], SystemMessage | UserMessage | AssistantMessage
    )


def _promote_prompt_result(result: list[Message] | UserContent) -> list[Message]:
    """Promote a prompt result to a list of messages.

    If the result is already a list of Messages, returns it as-is.
    If the result is UserContent, wraps it in a single user message.
    """
    if _is_messages(result):
        return result
    return [user(result)]


def _is_async_prompt(fn: Prompt | AsyncPrompt) -> TypeIs[AsyncPrompt]:
    return inspect.iscoroutinefunction(fn)


class PromptDecorator:
    """Protocol for the `prompt` decorator when used without a template."""

    @overload
    def __call__(
        self,
        fn: Prompt[P],
    ) -> MessagesPrompt[P]:
        """Decorator for creating prompts."""
        ...

    @overload
    def __call__(
        self,
        fn: AsyncPrompt[P],
    ) -> AsyncMessagesPrompt[P]:
        """Decorator for creating async prompts."""
        ...

    def __call__(
        self, fn: Prompt[P] | AsyncPrompt[P]
    ) -> MessagesPrompt[P] | AsyncMessagesPrompt[P]:
        """Decorator for creating a prompt."""
        if _is_async_prompt(fn):

            async def async_prompt(*args: P.args, **kwargs: P.kwargs) -> list[Message]:
                result = await fn(*args, **kwargs)
                return _promote_prompt_result(result)

            return async_prompt
        else:

            def prompt(*args: P.args, **kwargs: P.kwargs) -> list[Message]:
                result = fn(*args, **kwargs)
                return _promote_prompt_result(result)

            return prompt


class PromptTemplateDecorator:
    """Protocol for the `prompt` decorator when used with a template."""

    @overload
    def __call__(
        self,
        fn: Callable[P, None],
    ) -> MessagesPrompt[P]:
        """Decorator for creating prompts from template functions."""
        ...

    @overload
    def __call__(
        self,
        fn: Callable[P, Awaitable[None]],
    ) -> AsyncMessagesPrompt[P]:
        """Decorator for creating async prompts from template functions."""
        ...

    def __call__(
        self, fn: Callable[P, None] | Callable[P, Awaitable[None]]
    ) -> MessagesPrompt[P] | AsyncMessagesPrompt[P]:
        """Decorator for creating a prompt from a template function."""
        raise NotImplementedError()


@overload
def prompt(
    __fn: Prompt[P],
) -> MessagesPrompt[P]:
    """Create a decorator for sync Prompt functions (no arguments)."""
    ...


@overload
def prompt(
    __fn: AsyncPrompt[P],
) -> AsyncMessagesPrompt[P]:
    """Create a decorator for async Prompt functions (no arguments)."""
    ...


@overload
def prompt(
    *,
    template: None = None,
) -> PromptDecorator:
    """Create a decorator for Prompt functions (no template)"""


@overload
def prompt(
    *,
    template: str,
) -> PromptTemplateDecorator:
    """Create a decorator for template functions."""
    ...


def prompt(
    __fn: Prompt[P] | AsyncPrompt[P] | None = None,
    *,
    template: str | None = None,
) -> (
    MessagesPrompt[P]
    | AsyncMessagesPrompt[P]
    | PromptDecorator
    | PromptTemplateDecorator
):
    '''Prompt decorator for turning functions (or "Prompts") into prompts.

    This decorator transforms a function into a Prompt, i.e. a function that
    returns `list[llm.Message]`. Its behavior depends on whether it's called with a spec
    string.

    With a template string, it returns a PromptTemplateDecorator, in which case it uses
    the provided template to decorate an function with an empty body, and uses arguments
    to the function for variable substitution in the template. The resulting PromptTemplate
    returns messages based on the template.

    Without a template string, it returns a PromptFunctionalDecorator, which
    transforms a Prompt (a function returning either message content, or messages) into
    a PromptTemplate. The resulting prompt template either promotes the content into a
    list containing a single user message, or passes along the messages returned by the
    decorated function.

    Args:
        template: A string template with placeholders using `{{ variable_name }}`
            and optional role markers like [SYSTEM], [USER], and [ASSISTANT].

    Returns:
        A PromptTemplateDecorator or PromptFunctionalDecorator that converts
            the decorated function into a prompt.

    Spec substitution rules:
    - [USER], [ASSISTANT], [SYSTEM] demarcate the start of a new message with that role
    - [MESSAGES] indicates the next variable contains a list of messages to include
    - `{{ variable }}` injects the variable as a string, unless annotated
    - Annotations: `{{ variable:annotation }}` where annotation is one of:
        image, images, audio, audios, document, documents
    - Single content annotations (image, audio, document) expect a file path,
        URL, base64 string, or bytes, which becomes a content part with inferred mime-type
    - Multiple content annotations (images, audios, documents) expect a list
        of strings or bytes, each becoming a content part with inferred mime-type

    Examples:
        ```python
        @llm.prompt("""
            [SYSTEM] You are a helpful assistant specializing in {{ domain }}.
            [USER] {{ question }}
        """)
        def domain_question(domain: str, question: str) -> None:
            pass

        @llm.prompt()
        def answer_question(question: str) -> str:
            return f"Answer this question: {question}"
        ```
    '''
    if template:
        raise NotImplementedError()
    decorator = PromptDecorator()
    if __fn is None:
        return decorator
    return decorator(__fn)


class ContextPromptDecorator(Generic[P, DepsT]):
    """Protocol for the `context_prompt` decorator when used without a template."""

    @overload
    def __call__(
        self,
        fn: ContextPrompt[P, DepsT],
    ) -> ContextMessagesPrompt[P, DepsT]:
        """Decorator for creating context prompts."""
        ...

    @overload
    def __call__(
        self,
        fn: AsyncContextPrompt[P, DepsT],
    ) -> AsyncContextMessagesPrompt[P, DepsT]:
        """Decorator for creating async context prompts."""
        ...

    def __call__(
        self,
        fn: ContextPrompt[P, DepsT] | AsyncContextPrompt[P, DepsT],
    ) -> ContextMessagesPrompt[P, DepsT] | AsyncContextMessagesPrompt[P, DepsT]:
        """Decorator for creating a prompt."""
        raise NotImplementedError()


class ContextPromptTemplateDecorator(Protocol[DepsT]):
    """Protocol for the `context_prompt` decorator when used with a template."""

    @overload
    def __call__(
        self,
        fn: Callable[Concatenate[Context[DepsT], P], None],
    ) -> ContextMessagesPrompt[P, DepsT]:
        """Decorator for creating context prompts from template functions."""
        ...

    @overload
    def __call__(
        self,
        fn: Callable[Concatenate[Context[DepsT], P], Awaitable[None]],
    ) -> AsyncContextMessagesPrompt[P, DepsT]:
        """Decorator for creating async context prompts from template functions."""
        ...

    def __call__(
        self,
        fn: Callable[Concatenate[Context[DepsT], P], None]
        | Callable[Concatenate[Context[DepsT], P], Awaitable[None]],
    ) -> ContextMessagesPrompt[P, DepsT] | AsyncContextMessagesPrompt[P, DepsT]:
        """Decorator for creating a prompt from a template function."""
        raise NotImplementedError()


@overload
def context_prompt(
    __fn: ContextPrompt[P],
) -> ContextMessagesPrompt[P]:
    """Create a decorator for sync ContextPrompt functions (no arguments)."""
    ...


@overload
def context_prompt(
    __fn: AsyncContextPrompt[P],
) -> AsyncContextMessagesPrompt[P]:
    """Create a decorator for async ContextPrompt functions (no arguments)."""
    ...


@overload
def context_prompt(
    *,
    template: None = None,
) -> ContextPromptDecorator:
    """Create a decorator for ContextPrompt functions (no template)"""


@overload
def context_prompt(
    *,
    template: str,
) -> ContextPromptTemplateDecorator:
    """Create a decorator for template functions."""
    ...


def context_prompt(
    __fn: ContextPrompt[P] | AsyncContextPrompt[P] | None = None,
    *,
    template: str | None = None,
) -> (
    ContextMessagesPrompt[P]
    | AsyncContextMessagesPrompt[P]
    | ContextPromptDecorator
    | ContextPromptTemplateDecorator
):
    """Context Prompt decorator for turning functions (or "Prompts") into prompts."""
    raise NotImplementedError()
```

File: python/tests/llm/prompts/test_promt_decorator.py (hunk 3 above)
Comment (teamdandelion) [reply python/tests/llm/prompts/test_promt_decorator.py]: Ok, refactored as you suggested. And yeah, I included both sync and async for completeness in the "test the interface not the implementation" philosophy

File: python/mirascope/llm/prompts/decorator.py
Comment (teamdandelion) [reply python/mirascope/llm/prompts/decorator.py:1]: Used it, thanks!

File: python/tests/llm/prompts/test_promt_decorator.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        @llm.prompt
+        async def messages_prompt() -> list[llm.Message]:
+            return expected_messages
+
+        assert await messages_prompt() == expected_messages
+
+
+class TestValueErrorOnEmptyPrompts:
+    def test_sync_prompt_with_empty_list(
```
Comment (willbakst) [inline python/tests/llm/prompts/test_promt_decorator.py]: nit: why no parameterize these so we have one sync and one async test each with two parameters (that you return in the `empty_prompt` message)?

Comment (teamdandelion) [reply python/tests/llm/prompts/test_promt_decorator.py]: Redone as you suggested.

**Comment by teamdandelion** [comment]
### Merge activity

* **Aug 18, 7:33 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1242).
* **Aug 18, 7:33 PM UTC**: @teamdandelion merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1242).
